File: cartflow/src/app/services/cart-service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../environments/environment';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, tap, map, of, forkJoin} from 'rxjs';
import { AddToCartDto } from '../models/addToCart';
import { Product } from '../models/product';
import { CartItem } from '../models/cartItem';

@Injectable({
  providedIn: 'root',
})
export class CartService {
  private baseUrl = `${environment.apiUrl}/api/cart`;

  private cartCountSubject = new BehaviorSubject<number>(0);
  cartCount$ = this.cartCountSubject.asObservable();

  constructor(private http: HttpClient) {
    this.syncCountFromStorage();
  }

  private syncCountFromStorage() {
    const cart = JSON.parse(localStorage.getItem("cart") || '[]');
    const total = cart.reduce((sum: number, i: any) => sum + i.quantity, 0);
    this.cartCountSubject.next(total);
  }

  private updateCount(cart: any[]) {
    const total = cart.reduce((sum, i) => sum + i.quantity, 0);
    this.cartCountSubject.next(total);
  }

  // not logged In user can also add to cart 
  // cart saved in local storage 
  addToLocalCart(productId: number, quantity = 1) {
    const cart = JSON.parse(localStorage.getItem("cart") || '[]');
    const item = cart.find((x: any) => x.productId === productId);

    if (item) item.quantity += quantity;
    else cart.push({ productId, quantity });

    localStorage.setItem("cart", JSON.stringify(cart));
    this.updateCount(cart);
  }

  // clear cart 
  clearLocalCart() {
    localStorage.removeItem("cart");
    this.cartCountSubject.next(0);
  }

  // logged In User add to cart 
  addToDbCart(dto: AddToCartDto) {
    return this.http.post(`${this.baseUrl}/add`, dto).pipe(
      tap(() => {
        this.cartCountSubject.next(this.cartCountSubject.value + dto.quantity);
      })
    );
  }

  setDbCartCount(count: number) {
    this.cartCountSubject.next(count);
  }

  // get local cart 
   getLocalCart(): any[] {
    return JSON.parse(localStorage.getItem("cart") || '[]');
  }

  // get db cart 
  getDbCart() {
    return this.http.get<CartItem[]>(`${this.baseUrl}`);
  }

  removeFromDbCart(productId: number) {
  return this.http
    .delete(`${this.baseUrl}/${productId}`)
    .pipe(
      tap(() => {
        // decrement count by 1 (or more if you support quantities later)
        const current = this.cartCountSubject.value;
        this.cartCountSubject.next(Math.max(current - 1, 0));
      })
    );
}


    hydrateLocalCart() {
        const raw = this.getLocalCart();

        if (!raw.length) {
          return of([]);
        }

      return forkJoin(
        raw.map(item => 
          this.http.get<Product>(`${environment.apiUrl}/api/product/${item.productId}`).pipe(
            map(product => ({
              productId: product.id,
              name: product.name,
              price: product.price,
              imageUrl: product.imageUrl,
              quantity: item.quantity
            }))
          )
        )
      );
  }

 mergeCart(items: AddToCartDto[]) {
  return this.http.post(`${this.baseUrl}/merge`, items).pipe(
    tap(() => {
      const total = items.reduce((s, i) => s + i.quantity, 0);
      this.cartCountSubject.next(total);
    })
  );
}



}


// public async Task CreateOrderAsync(int userId, CreateOrderDto dto)
// {
//     var cartItems = await _cartRepo.GetUserCartAsync(userId);
//     if (!cartItems.Any())
//         throw new InvalidOperationException("Cart is empty");

//     var order = new Order
//     {
//         UserId = userId,
//         Status = "Pending",
//         CreatedAt = DateTime.UtcNow,
//         Total = cartItems.Sum(i => i.Quantity * i.Product.Price),
//         Items = cartItems.Select(i => new OrderItem
//         {
//             ProductId = i.ProductId,
//             ProductName = i.Product.Name,
//             PriceAtPurchase = i.Product.Price,
//             Quantity = i.Quantity
//         }).ToList()
//     };

//     await _orderRepo.AddAsync(order);
//     await _cartRepo.ClearCartAsync(userId);
// }

File: backend/Services/CartService.cs
using CartFlow.DTOs;
using CartFlow.Entities;
using CartFlow.Interfaces;

namespace CartFlow.Services
{
    public class CartService : ICartService
    {
        private readonly ICartRepository _cartRepo;
        private readonly IProductRepository _productRepo;

        public CartService(ICartRepository cartRepo, IProductRepository productRepo)
        {
            _cartRepo = cartRepo;
            _productRepo = productRepo;
        }

        // Return user's cart items mapped to CartItemDto
        public async Task<IEnumerable<CartItemDto>> GetCartByUserIdAsync(int userId)
        {
            var items = await _cartRepo.GetUserCartAsync(userId);
            var result = new List<CartItemDto>();

            foreach (var i in items)
            {
                result.Add(new CartItemDto
                {
                    Id = i.Id,
                    ProductId = i.ProductId,
                    Name = i.Product?.Name ?? string.Empty,
                    ImageUrl = i.Product?.ImageUrl ?? string.Empty,
                    Price = i.Product?.Price ?? 0,
                    Quantity = i.Quantity
                });
            }

            return result;
        }

        // Add to cart
        public async Task AddToCartAsync(int userId, AddToCartDto dto)
        {
            // validate product exists
            var p = await _productRepo.GetByIdAsync(dto.ProductId);
            if (p == null) throw new InvalidOperationException("Product not found.");

            var existing = await _cartRepo.GetCartItemAsync(userId, dto.ProductId);
            if (existing == null)
            {
                var item = new CartItem
                {
                    UserId = userId,
                    ProductId = dto.ProductId,
                    Quantity = dto.Quantity
                };
                await _cartRepo.AddAsync(item);
                return;
            }

            // if exists, increase quantity
            existing.Quantity += dto.Quantity;
            await _cartRepo.UpdateAsync(existing);
        }

        //merge cart 
        public async Task MergeCartAsync(int userId, List<AddToCartDto> items)
        {
            foreach (var dto in items)
            {
                var existing = await _cartRepo.GetCartItemAsync(userId, dto.ProductId);

                if (existing == null)
                {
                    var item = new CartItem
                    {
                        UserId = userId,
                        ProductId = dto.ProductId,
                        Quantity = dto.Quantity
                    };
                    await _cartRepo.AddAsync(item);
                }
                else
                {
                    existing.Quantity += dto.Quantity;
                    await _cartRepo.UpdateAsync(existing);
                }
            }
        }


        // update cart quantity
        public async Task UpdateQuantityAsync(int userId, int productId, int quantity)
        {
            if (quantity <= 0) throw new ArgumentException("Quantity must be > 0");

            var item = await _cartRepo.GetCartItemAsync(userId, productId);
            if (item == null) throw new InvalidOperationException("Cart item not found.");

            item.Quantity = quantity;
            await _cartRepo.UpdateAsync(item);
        }

        // remove product from cart
        public async Task RemoveFromCartAsync(int userId, int productId)
        {
            var item = await _cartRepo.GetCartItemAsync(userId, productId);
            if (item == null) return;
            await _cartRepo.RemoveAsync(item);
        }

        public async Task ClearCartAsync(int userId)
        {
            await _cartRepo.ClearUserCartAsync(userId);
        }
    }
}
